op-node/rollup/derive/channel_out.go

```diff
@@ -99,40 +125,119 @@ func (co *ChannelOut) AddBlock(block *types.Block) (uint64, error) {
+	case SpanBatchType:
+		return co.writeSpanBatch(batch)
+	default:
+		return 0, fmt.Errorf("unrecognized batch type: %d", co.batchType)
+	}
+}
+
+func (co *ChannelOut) writeSingularBatch(batch *SingularBatch) (uint64, error) {
+	var buf bytes.Buffer
 	// We encode to a temporary buffer to determine the encoded length to
 	// ensure that the total size of all RLP elements is less than or equal to MAX_RLP_BYTES_PER_CHANNEL
-	var buf bytes.Buffer
-	if err := rlp.Encode(&buf, batch); err != nil {
+	if err := rlp.Encode(&buf, NewSingularBatchData(*batch)); err != nil {
 		return 0, err
 	}
 	if co.rlpLength+buf.Len() > MaxRLPBytesPerChannel {
-		return 0, fmt.Errorf("could not add %d bytes to channel of %d bytes, max is %d. err: %w",
+		co.fullErr = fmt.Errorf("could not add %d bytes to channel of %d bytes, max is %d. err: %w",
 			buf.Len(), co.rlpLength, MaxRLPBytesPerChannel, ErrTooManyRLPBytes)
+		return 0, co.fullErr
 	}
 	co.rlpLength += buf.Len()
 
 	// avoid using io.Copy here, because we need all or nothing
 	written, err := co.compress.Write(buf.Bytes())
+	if co.compress.FullErr() != nil {
+		co.fullErr = co.compress.FullErr()
+	}
+	return uint64(written), err
+}
+
+// writeSpanBatch appends a SingularBatch to the channel's SpanBatch.
+// A channel can have only one SpanBatch. And compressed results should not be accessible until the channel is closed, since the prefix can be changed.
+// So it resets channel contents and rewrites the entire SpanBatch each time, and compressed results are copied to reader after the channel is closed.
+// It makes we can only get frames once the channel is full or closed, in the case of SpanBatch.
+func (co *ChannelOut) writeSpanBatch(batch *SingularBatch) (uint64, error) {
+	if co.fullErr != nil {
+		// channel is already full
+		return 0, co.fullErr
+	}
+	var buf bytes.Buffer
+	// Append Singular batch to its span batch builder
+	co.spanBatchBuilder.AppendSingularBatch(batch)
+	// Convert Span batch to RawSpanBatch
+	rawSpanBatch, err := co.spanBatchBuilder.GetRawSpanBatch()
+	if err != nil {
+		return 0, err
+	}
+	// Encode RawSpanBatch into bytes
+	if err := rlp.Encode(&buf, NewSpanBatchData(*rawSpanBatch)); err != nil {
+		return 0, err
+	}
+	co.rlpLength = 0
+	// Ensure that the total size of all RLP elements is less than or equal to MAX_RLP_BYTES_PER_CHANNEL
+	if co.rlpLength+buf.Len() > MaxRLPBytesPerChannel {
+		co.fullErr = fmt.Errorf("could not add %d bytes to channel of %d bytes, max is %d. err: %w",
+			buf.Len(), co.rlpLength, MaxRLPBytesPerChannel, ErrTooManyRLPBytes)
+		return 0, co.fullErr
+	}
+	co.rlpLength = buf.Len()
+
+	if co.spanBatchBuilder.GetBlockCount() > 1 {
+		// Flush compressed data into reader to preserve current result.
+		// If the channel is full after this block is appended, we should use preserved data.
+		if err := co.compress.Flush(); err != nil {
+			return 0, err
+		}
+		_, err := io.Copy(co.reader, co.compress)
+		if err != nil {
+			return 0, err
```

Comment:
I think we should return `written` to follow the convention in `writeSingularBatch`. If we write to `reader`, its size must be returned.

Reply:
Use Write()